NuGet/NuGet.Client
# Emit ETW intervals for better performance tracing of PM UI User Gestures

## Bug



Fixes: https://github.com/NuGet/Client.Engineering/issues/1245

Regression? Last working version:

## Description


### Summary:
- Using the Visual Studio Common ETW provider, emit start/stop events from every "user gesture" (eg, button click, checkbox change, selection change, search text typed).
  - Applied to 12 PM UI events (see below).
- Consuming these ETW events can be done via VS Feedback tickets where ETL files are attached. We could at least have some idea which area within PM UI appears to be unusually slow.
- Debugging NuGet.Client performance issues is another use-case for these events.

### Refactor IntervalTracker
- Added support for a tracker which does not maintain all intervals. 
- The existing tracker always accumulated intervals and emitted them as a telemetry event later. I'm not doing that with these PM UI events (for now).

### PM UI Events affected:

- **PackageManagerLoadedAsync**
- **PackageList_SelectionChanged** - select a package in packages list
- **Filter_SelectionChanged** - change Browse/Installed/Updates/Consolidate tabs
- **NuGetPackageManagerControlSearchTask** - type into the search box
- **ClearSearch** - Use the search box's clear ("X ") button
- **ExecuteRestartSearchCommand** - Press the Refresh button
- **OnProjectChanged** - affects Solution PMUI only; add/remove or load/unload projects to your solution
- **OnProjectActionsExecuted** - install/update/uninstall a package using PM UI
- **OnNuGetCacheUpdated** - fired when a project nomination happens
- **Sources_PackageSourcesChanged** - changed package source
- **SourceRepoList_SelectionChanged** - The repository list changed (that means updated in the tools -> Options -> NuGet UI)
- **CheckboxPrerelease_CheckChanged** - The checkbox for prerelease has been checked

### How to test using PerfView 

1. The following command will launch PerfView, immediately begin collecting events, focused on devenv.exe, and with the VisualStudio Common provider enabled.

`PerfView64.exe "/DataFile:PerfViewData.etl" /BufferSizeMB:256 /StackCompression /CircularMB:500 /KernelEvents:ThreadTime /TplEvents:Default -Providers:641d7f6c-481c-42e8-ab7e-d18dc5e5cb9e,*Microsoft-VisualStudio-Common /FocusProcess:"devenv.exe" collect`

2. Once you've executed your actions of interest, press "Stop collection" on the PerfView window.

3. PerfView will build a .zip file, this can take a minute or two depending on how long you recorded.

4. Open the etl.zip file within PerfView. 
5. Open the "Events" section:
![image](https://user-images.githubusercontent.com/49205731/136678325-4497e675-166f-4359-a7c1-2558c2ccf99f.png)

6. In the Events window, type "pmui" in the Event Types Filter textbox.

7. Select each event type (or ctrl+a to select all) then press Enter to view events for those type(s).

8. You should see all PM UI events recorded in this trace.
Example: 
"**Microsoft-VisualStudio-Common/vs_nuget_pmuigesture_filter_selectionchanged_installed/Start**"

All "**/Stop**" events will contain the elapsed time in milliseconds under DURATION_MSEC.

### Example Analysis: Installing a package
Installing a package, I see that the event time is about 14% less than the Output Window's "Time Elapsed".
![image](https://user-images.githubusercontent.com/49205731/136678355-f5bd2baa-7474-480f-a9c0-89f4670c0d78.png)

### Another Example with many events
![image](https://user-images.githubusercontent.com/49205731/136678296-27ca8b43-4fbe-41cc-a929-271441625afb.png)


## PR Checklist


- **Tests**
  - [x] Automated tests added - IntervalTracker modifications are tested.
  - **OR**
  
  - [x] Test exception - New events are not tested with automation. To test, use PerfView, execute an action, and ensure a corresponding event is emitted.
  - **OR**

- **Documentation**
  
  - **OR**
